applications/finetune-quora-embeddings/finetune.py

```diff
@@ -0,0 +1,143 @@
+from modal import Image, gpu, Stub, Volume, Secret
+
+
+# Finetune Configuration
+MAXIMUM_TRAIN_DATASET_USAGE = 0.1
+
+# Stub Configuration
+MODEL_ID = "BAAI/bge-base-en-v1.5"
+GPU_CONFIG = gpu.A10G()
+
+stub = Stub("finetune")
+
+
+def download_model_and_dataset():
+    from sentence_transformers import SentenceTransformer
+    from datasets import load_dataset
+
+    SentenceTransformer(MODEL_ID)
+    load_dataset("quora", "default", "train", num_proc=6)
+
+
+image = (
+    Image.debian_slim()
+    .pip_install("sentence-transformers", "torch", "datasets", "scikit-learn", "tqdm")
+    .run_function(download_model_and_dataset)
+)
+
+
+def generate_quora_input_example(examples):
+    from sentence_transformers import InputExample
+
+    return [
+        InputExample(
+            texts=[
+                example["questions"]["text"][0],
+                example["questions"]["text"][1],
+            ],
+            label=int(example["is_duplicate"]),
+        )
+        for example in examples
+    ]
+
+
+@stub.function(image=image, gpu=GPU_CONFIG, timeout=3600)
+def train_model(train_dataset_percentage: float):
+    from sentence_transformers import SentenceTransformer, losses, evaluation, util
+    from sklearn.metrics import roc_auc_score
+    from torch.utils.data import DataLoader
+    from datasets import load_dataset
+    import math
+    import time
+    from tqdm import tqdm
+    import numpy as np
+
+    # Define the model. Either from scratch of by loading a pre-trained model
+    model = SentenceTransformer(MODEL_ID)
+    dataset = load_dataset("quora", "default", "train")["train"]
+    # Define your train examples. You need more than just two examples...
+    train_test_split = dataset.train_test_split(test_size=0.1, seed=42)
+    dataset_slice = math.floor(
+        len(train_test_split["train"]) * train_dataset_percentage
+    )
+    train_dataset = generate_quora_input_example(
+        train_test_split["train"].select(range(dataset_slice))
+    )
+
+    test_dataset = generate_quora_input_example(train_test_split["test"])
+
+    # Define your train dataset, the dataloader and the train loss
+    train_dataloader = DataLoader(train_dataset, shuffle=True, batch_size=16)
+    train_loss = losses.OnlineContrastiveLoss(model)
+    evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(
+        test_dataset,
+    )
+
+    loss = evaluator(model)
+
+    # for i in range(2):
```

Comment:
commented out because was working on the optimising of the loop ( ~ 40k elements in test set so was taking up a lot of time )